File: solutions/python/binary-search/kth-missing-positive-number.py
# LeetCode 1539 — Kth Missing Positive Number (Easy)
# Category: Binary Search · Approach: Binary Search
# Time: O(log n) | Space: O(1)
# Source: https://leetcode.com/problems/kth-missing-positive-number/

def find_kth_positive(arr: list[int], k: int) -> int:
    lo, hi = 0, len(arr)
    while lo < hi:
        mid = (lo + hi) // 2
        if arr[mid] - (mid + 1) < k:
            lo = mid + 1
        else:
            hi = mid
    return lo + k

File: solutions/typescript/binary-search/kth-missing-positive-number.ts
// LeetCode 1539 — Kth Missing Positive Number (Easy)
// Category: Binary Search · Approach: Binary Search
// Time: O(log n) | Space: O(1)
// Source: https://leetcode.com/problems/kth-missing-positive-number/

function findKthPositive(arr: number[], k: number): number {
  let lo = 0;
  let hi = arr.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (arr[mid] - (mid + 1) < k) lo = mid + 1;
    else hi = mid;
  }
  return lo + k;
}
